src/playground-code-editor.ts

```diff
@@ -271,7 +289,33 @@ export class PlaygroundCodeEditor extends LitElement {
       >;
       for (const prop of changedTyped.keys()) {
         switch (prop) {
+          case 'documentKey': {
+            if (!this.documentKey) {
+              break;
+            }
+            let docInstance = this._privateDocCache.get(this.documentKey);
+            if (!docInstance) {
+              docInstance = new CodeMirror.Doc(this.value ?? '');
+              this._privateDocCache.set(this.documentKey, docInstance);
+            } else {
+              if (this.value && docInstance.getValue() !== this.value) {
+                // A `documentKey` and `value` have both been provided and the
+                // cached document instance doesn't match the passed in value.
+                // Set the new `value` on the document instance.
+                docInstance.setValue(this.value ?? '');
+              }
+            }
+            this._valueChangingFromOutside = true;
+            cm.swapDoc(docInstance);
+            this._valueChangingFromOutside = false;
+            break;
+          }
           case 'value':
+            if (changedTyped.has('documentKey')) {
+              // We've already processed the `documentKey` and don't want to
```

Comment:
How about the case where all of these are true in the same `update()`: (1) We didn't have a  `documentKey` before, (2) A `documentKey` is set, (3) A `value` is set. It seems like we might ignore the new `value` right now, because both the `documentKey` and the `value` switch cases `break` early in that case?

I wonder if it would help to hoist the `documentKey` handling above the `switch` statement all together, so that we can guarantee we always do the document swap before we look at any new `value`. Then the `value` case can probably just remain as it was? (To make type checking happy we could just have a no-op `case` for `documentKey` with a brief comment saying it was handled above).

Reply:
If a value is on `hasChanged` it's already updated on the instance.
In the given case (and I think it's verified by tests), `documentKey` has changed so we will skip the `value` block. And in the `documentKey` block we instantiate a new document instance from `this.value`  and swap it into codemirror which populates the editor.

Another reason `documentKey` must handle `value` cases is that setting values add history.

However you've raised the case where `documentKey` is unset to undefined. That case would indeed break in both places, so I'll add a test and handling to it.

Reply:
Addressed by removing the `documentKey` undefined case, and instead have a fallback default documentKey associated with the playground editor instance.

Reply:
Cool!

Just a thought -- maybe a way of simplifying the implementation for handling the `undefined` case could be to have an internal-only cache key that represents "default" documents. That could give you less special casing.

Also I was wondering about what would happen in this sequence: (1) type something (2) set document key to something (3) type something (4) set document key to undefined (5) undo. Would (5) undo the thing typed in (1)? I would think it probably shouldn't, which implies that the "default document" should get reset each time the document key gets set to undefined.

Reply:
Oh sorry, I see I sent this just as you sent your latest commit :)